Add airport styling

AnonymousAlligatorOSM opened the pull request:
Add styling for airports. See #188.

## Zoomed out
At zoom level <=10 and when the long name doesn't fit, the IATA or ICAO code is shown.

![Screenshot from 2022-05-08 22-11-24](https://user-images.githubusercontent.com/101657206/167334869-32ca5b53-eb3a-494d-b23f-da71e2e108ce.png)

## Large airport
![Screenshot from 2022-05-08 22-11-44](https://user-images.githubusercontent.com/101657206/167334905-40b5f5f1-0924-4e92-97c5-2351edb6ad41.png)

## Smaller airport
![Screenshot from 2022-05-08 22-12-12](https://user-images.githubusercontent.com/101657206/167334928-d2058ac3-fe00-47b0-86f0-b67680066df3.png)

## Very small airport
![Screenshot from 2022-05-08 22-20-23](https://user-images.githubusercontent.com/101657206/167335082-998088b2-68c5-4fc9-bdc2-94863410c889.png)


## Gate numbers
Bonus points if you recognize this airport

![Screenshot from 2022-05-08 22-12-51](https://user-images.githubusercontent.com/101657206/167334951-ff4ea71a-243e-4db8-8d85-b6261d22de5d.png)

# Remaining Issues
- Military air bases are not styled differently--see #232.
- Should international airports stand out more somehow?

claysmalley commented:
I think the outline without runways looks a bit empty. How about having the runways show up at the same zoom as the outline? They're pretty big features and I think they'd look good at low zoom.

Likewise, taxiways show up at a somewhat high zoom and airports look empty next to residential streets. What if we bump taxiways down to where runways are currently showing up?

![Screenshot from 2022-05-09 00-00-39](https://user-images.githubusercontent.com/1732117/167338460-ce7b0030-d9e0-4b32-95c2-651bc1b85570.png) ![Screenshot from 2022-05-09 00-00-15](https://user-images.githubusercontent.com/1732117/167338457-4c0988d3-2c5b-46db-bd9e-190fc4f9f1ae.png) ![Screenshot from 2022-05-09 00-00-26](https://user-images.githubusercontent.com/1732117/167338459-b1c4bbdd-e61a-4e53-911a-a0379201b755.png)

claysmalley commented:
I do like how everything looks overall, though! Fantastic work.

AnonymousAlligatorOSM commented:
Agreed, I think that looks good.

wmisener commented:
I really like the use of IATA codes when zoomed out, and I think that combined with the runways will make it clear these represent airports at low zoom, even without an airplane icon. I just want to point out that this is the first non-car transport-related rendering, and also the first use of purple on the map, so this might set a precedent for purple being a transportation-related color (similar to Carto). Not that this is a bad thing (I think it looks good), but something to consider going forward, especially if purple was being considered for any other large features.

claysmalley commented:
I like the stroke width and fill transparency, but a different color would help distinguish aeroways from boundaries. The shades of purple in this PR are very close to the ones used for administrative boundaries, which can interact with runways and taxiways confusingly:
![Screenshot from 2022-05-09 09-20-08](https://user-images.githubusercontent.com/1732117/167418864-f5a8ec0c-da5f-4914-b5bc-ea32c38c38e6.png) ![Screenshot from 2022-05-09 09-18-40](https://user-images.githubusercontent.com/1732117/167418670-f7cb7627-3ec9-4a46-8d01-77f753cb7f37.png)

AnonymousAlligatorOSM commented:
Changed to a more pink color to avoid confusion with boundaries:
![Screenshot from 2022-05-09 10-36-29](https://user-images.githubusercontent.com/101657206/167445942-2d21dabd-6b06-4f56-ae6c-a35e467474ce.png)

And made airport labels without IATA and ICAO codes not show up until higher zoom levels:
![Screenshot from 2022-05-09 10-37-55](https://user-images.githubusercontent.com/101657206/167446013-327e4c3f-0a61-49b7-bb58-ddcce5b7ba72.png)

OpenMapTiles doesn't expose the runway surface tag.

ZeLonewolf commented:
The pink doesn't feel quite right to me.  I think the original purple shading was better, but there may be other colors that work as well.

The intersection of boundaries and airports feels like a bit of an esoteric case to me.  Perhaps we can monkey with the draw order and not necessarily show all parts of the boundary drawing.  I'm not sure why a map user would care about the precise position of a town boundary through an airport anyways.

I'm also thinking that when we're down to the IATA codes that it should have an airplane icon, though I'm happy to defer that to a future issue if the author isn't graphically inclined.

adamfranco commented:
I'd have to agree with Brian that the purple "looked better". This style already uses a lot of red tones and I'm wondering if a red tone would also be appropriate for military areas. I like the idea of using a consistent color-range for public transit broadly (like busways, airports, ferry terminals, train & bus stations). I don't know that choosing a public-transport color needs to block this PR, but maybe we can take an initial stab (purple or otherwise) and then shift it later as more features get added.

adamfranco commented:
I just noticed that highway tunnels are rendering on top of runways and taxi-ways ([LAV on localhost](http://localhost:1776/#13.79/36.08308/-115.15235), [ORD on localhost](http://localhost:1776/#16.61/41.958981/-87.927631)):
![Screen Shot 2022-05-09 at 12 01 15 PM](https://user-images.githubusercontent.com/25242/167450302-33cef97d-c27e-41a4-b610-07db52677b66.png)

These lines should be moved after the highway tunnels:
```
  lyrAeroway.runway,
  lyrAeroway.runwayArea,
  lyrAeroway.taxiway,
```

AnonymousAlligatorOSM commented:
Reverted the color and added icons:
![Screenshot from 2022-05-09 11-31-10](https://user-images.githubusercontent.com/101657206/167455622-59f5773b-5780-4368-b708-8f0db6f879a1.png)

ZeLonewolf commented:
Looks like the only question is boundary collisions.  How would people feel about putting airport areas on top of boundaries?

Also bear in mind that at some point I anticipate putting labels on boundaries to indicate what territory is on either side.  This would be awfully cluttered to have a bunch of boundary labels running through ATL.

claysmalley commented:
France, fix your administrative boundaries...
![Screenshot from 2022-05-09 18-49-40](https://user-images.githubusercontent.com/1732117/167513326-556853e5-be3b-47a8-b79c-de2e6c9a10a8.png)

claysmalley commented:
In all seriousness, how do people feel about a more skeuomorphic runway style, with a thick black line and a thin dashed white line down the middle?

ZeLonewolf commented:
>  how do people feel about a more skeuomorphic runway style, with a thick black line and a thin dashed white line down the middle?

If we do that for runways it should be done for roads also, and IMO that would only really make sense at the highest zooms.

zekefarwell commented:
Although I like the skumorphic idea, I think it would make more sense in a map style going for a realistic look and feel.   The highway map aesthetic we are going for in this style.is more abstract and symbolic.

zekefarwell commented:
I like the look of this so far.  Definitely also a fan of the cooler purple over the warmer pink.  However, a boundary running through an airport does look rather muddled with the similar shade of purple.  I wonder if desaturating the purple further or even moving entirely to a shade of gray would work.

I noticed that the North Carolina state highway map shows the runway structure of airports in the city insets.  White with a black outline is used.  Some examples below.  If anyone else has maps that show airport details like this, please share.
![PXL_20220510_004955769](https://user-images.githubusercontent.com/281482/167522591-3a8225e0-4bf1-4ddf-88c0-d027b7acae74.jpg)
![PXL_20220510_004915950](https://user-images.githubusercontent.com/281482/167522592-51c3a424-f682-4985-b3ab-e040ca6b652c.jpg)
![PXL_20220510_004735699](https://user-images.githubusercontent.com/281482/167522595-5af4212d-1e14-4b49-a8dc-67a442c743b4.jpg)

ZeLonewolf commented on style/constants/color.js:
```diff
@@ -12,3 +12,8 @@ export const borderCasing = `hsl(${hueBorderCasing}, 35%, 86%)`;
 export const parkOutline = "hsl(136, 41%, 79%)";
 export const parkLabel = "hsl(136, 71%, 29%)";
+
+export const airportFill = "hsla(292, 41%, 89%, .3)";
```
Alpha blend will be problematic because the color will be different depending on what's behind.  I used the color picker, and the equivalent hsl is `hsl(333, 33%, 93%)`

ZeLonewolf commented:
Exploring grayscale as suggested:

![image](https://user-images.githubusercontent.com/3254090/168198107-4d27990e-24cd-4f54-ab92-0fd1561a97fb.png)

ZeLonewolf commented:
I played around with changing the layer ordering to see what we could do about airfield/boundary collisions.  The results weren't great.  First, I tried moving the boundary casing below the airfield fill.  This had the effect of keeping the boundaries on top of the airfield, but making them incoherent as you lose the state/county/city distinctions.  I also tried moving the boundary casing and fill below the airfield, and this kept the airfield intact no problem but it creates a love triangle between airfield, water, and boundary.  Airfield wants to be above boundary, boundary wants to be above water, and water wants to be above airfield (for example, small retaining ponds and so forth).  And that's the ball game.

I think the bottom line here is that having a boundary run through an airport is obviously stupid (in reality) and therefore this stupidity is reflected, accurately, on the map.  So we probably have to live with those collisions unless someone has a magic sauce solution here.

I'd propose that we come to a consensus on what color scheme we like and push ahead.

claysmalley commented:
What about grayscale runways/taxiways and purple airport boundaries? I like the grayscale runways, and maybe that alone is enough to make them conflict with administrative boundaries less.

jleedev commented:
More distinguishable from boundaries, harder to tell what's a road
![Screenshot_20220513-125408](https://user-images.githubusercontent.com/23022/168331911-b8c49ed0-4d85-4683-9683-bc51abf56ace.png)
![Screenshot_20220513-125447](https://user-images.githubusercontent.com/23022/168331913-96dc0e38-448b-4ec6-8d04-8438a145a791.png)

ZeLonewolf commented:
Of note, boundaries on top of airports is silly on osm-carto as well :-D
https://www.openstreetmap.org/search?query=CDG%20airport#map=14/49.0135/2.5557

ZeLonewolf commented:
The main issue with the screenshot in https://github.com/ZeLonewolf/openstreetmap-americana/pull/318#issuecomment-1126258309 is that we chose to put tertiary roads in gray, hence the conflict.  If we could undo that, we might have this nailed.

zekefarwell commented on style/constants/color.js:
```diff
@@ -12,3 +12,8 @@ export const borderCasing = `hsl(${hueBorderCasing}, 35%, 86%)`;
 export const parkOutline = "hsl(136, 41%, 79%)";
 export const parkLabel = "hsl(136, 71%, 29%)";
+
+export const airportFill = "hsla(292, 41%, 89%, .3)";
+export const airportOutline = "hsl(292, 41%, 79%)";
+export const airportRunway = "hsl(270, 41%, 79%)";
+export const airportLabel = "hsl(270, 71%, 29%)";
```
I think shifting the color hue more towards more of a bluish-purple makes the warmer purple boundary line more distinct while also leaving a distinction between gray roads and bluish-purple taxiways.

![Screen Shot 2022-05-13 at 11 10 38 PM](https://user-images.githubusercontent.com/281482/168408795-4703f0fd-3c80-4e5e-9a67-1656bae621b8.png)
![Screen Shot 2022-05-13 at 11 10 57 PM](https://user-images.githubusercontent.com/281482/168408798-7041b9db-9bc7-417e-9b73-3d1e4bb42c46.png)

Here's the code change:

```suggestion
export const airportFill = "hsl(250, 41%, 95%)";
export const airportOutline = "hsl(250, 41%, 79%)";
export const airportRunway = "hsl(250, 41%, 79%)";
export const airportLabel = "hsl(250, 71%, 29%)";
```

Moving the dotted boundary line (but not the purple casing) above everything else (except labels) also is an improvement I'd say, but this should probably be considered in a different PR to keep this one focused on airports.

![Screen Shot 2022-05-13 at 11 09 55 PM](https://user-images.githubusercontent.com/281482/168408966-89d30ea5-47f6-4e30-8c0a-a5109dd70e0e.png)

claysmalley commented on style/layer/aeroway.js:
```diff
@@ -0,0 +1,250 @@
+    },
+  },
+  layout: {
+    "line-cap": "round",
```
Try using `square` instead of `round` on line caps for runways and taxiways. Since these features tend to have square ends in real life, I think they look a little better this way.

ZeLonewolf commented:
Go team!

1. I really like the color work from @zekefarwell in https://github.com/ZeLonewolf/openstreetmap-americana/pull/318#discussion_r872921754.  I think we have a winner there unless anyone objects.
2. I also like the skeuomorphic runway lines that @claysmalley demonstrated, but only at the highest zoom where it's clear that we're looking at dashed lines.  But, we can do that as a separate PR
3. I agree with the suggestion by @claysmalley for square end caps https://github.com/ZeLonewolf/openstreetmap-americana/pull/318#discussion_r872984695

Thanks for spearheading this @AnonymousAlligatorOSM, this is the first significant small-area feature and it will set the stage for other area features.

ZeLonewolf reviewed (approved):
I think this is ready to ship in this state.  We might consider some improvements at zoom 10 as far as which features to show for what types of airports, but I think those are tweaks that can be separately investigated.

claysmalley reviewed (approved):
Looks good to me! Thank you for working on this.

adamfranco commented:
I'm away from a computer and can't check myself, but just be sure that
there runway dashes don't conflict with unpaved roads. I can't quite tell
from the screenshots posted if that is an issue or not.

On Sat, May 14, 2022, 7:36 PM Brian Sperlongano ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> I think this is ready to ship in this state. We might consider some
> improvements at zoom 10 as far as which features to show for what types of
> airports, but I think those are tweaks that can be separately investigated.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ZeLonewolf/openstreetmap-americana/pull/318#pullrequestreview-973107082>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAGFGT27ZBAF4MONIRI3QLVKA2HBANCNFSM5VM54PUQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

ZeLonewolf commented:
@adamfranco the runway dashes are not included in this PR.  @claysmalley only posted samples.
